### PR TITLE
Adding overwrite or ignore behavior for row conflicts.

### LINKED ISSUE #14553: Adding (Insert or update if key exists)  option to `.to_sql`
Suppose you have an existing SQL table called `person_age`, where `id` is the primary key:

```python

    age
id	
1	18
2	42
```

and you also have new data in a `DataFrame` called `extra_data`

```python
	age
id	
2	44
3	95
```

then it would be useful to have an option on `extra_data.to_sql()` that allows to pass the DataFrame to SQL with an `INSERT` or `UPDATE` option on the rows, based on the `primary key`.

In this case, the `id=2` row would get updated to  `age=44` and the `id=3` row would get added 

#### Expected Output

```python
	age
id	
1	18
2	44
3	95
```

#### (Maybe) helpful code references

+ [Use `merge` from SQLAlchemy](http://stackoverflow.com/questions/7889183/sqlalchemy-insert-or-update-example)?
+ The query: '''INSERT or REPLACE into person_age (id, age) values (?,?,?) ''' in [this question](http://stackoverflow.com/questions/23574614/appending-pandas-dataframe-to-sqlite-table-by-primary-key)

I looked at `pandas` `sql.py` sourcecode to come up with a solution, but I couldn't follow.

#### Code to replicate the example above

(Apologies for mixing `sqlalchemy` and `sqlite`

```python
import pandas as pd
from sqlalchemy import create_engine
import sqlite3
conn = sqlite3.connect('example.db')
 
c = conn.cursor()
c.execute('''DROP TABLE IF EXISTS person_age;''')
c.execute('''
          CREATE TABLE person_age
          (id INTEGER PRIMARY KEY ASC, age INTEGER NOT NULL)
          ''')
conn.commit()
conn.close()

##### Create original table

engine = create_engine("sqlite:///example.db")
sql_df = pd.DataFrame({'id' : [1, 2], 'age' : [18, 42]})

sql_df.to_sql('person_age', engine, if_exists='append', index=False)


#### Extra data to insert/update

extra_data = pd.DataFrame({'id' : [2, 3], 'age' : [44, 95]})
extra_data.set_index('id', inplace=True)

#### extra_data.to_sql()  with row update or insert option

expected_df = pd.DataFrame({'id': [1, 2, 3], 'age': [18, 44, 95]})
expected_df.set_index('id', inplace=True)
```

### PR DESCRIPTION
**Pull request by LSturtew**
@cvonsteg started the effort to implement this. Since they are unable to continue and this is a highly desired feature, I made an attempt to clean up the code and fix merge conflicts. I would like to share the credit for this PR (I do not know how this works).

A new parameter is added to `to_sql`: `on_row_conflct`

    on_row_conflict : {'fail', 'overwrite', 'append'},
        default: 'fail'
        Determine insertion behavior in case of a primary key clash.
        - 'fail': Do nothing to handle primary key clashes, will raise an Error.
        - 'overwrite': Update existing rows in database with primary key clashes,
          and append the remaining rows with non-conflicting primary keys
        - 'append': Ignore incoming rows with primary key clashes, and
          insert only the incoming rows with non-conflicting primary keys

- [x] closes #14553, continuation of #29636
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -2799,13 +2800,22 @@ def to_sql(
         schema : str, optional
             Specify the schema (if database flavor supports this). If None, use
             default schema.
-        if_exists : {'fail', 'replace', 'append'}, default 'fail'
+        if_exists : {'fail', 'replace', 'append'},\
+        default 'fail'
```
Comment (jreback) [inline pandas/core/generic.py:2804]: why are you u using row-continuation here? just leave it

can consider deprecating this and change the name to `if_table_exists` to be more obvious (but orthogonal)

File: pandas/core/generic.py [hunk 2]
```diff
@@ -2799,13 +2800,22 @@ def to_sql(
...
             * fail: Raise a ValueError.
             * replace: Drop the table before inserting new values.
             * append: Insert new values to the existing table.
 
+        on_row_conflict : {'fail', 'overwrite', 'ignore'},\
+        default 'fail'
+            How to behave if a row already exists.
+
+            * fail: Raise ValueError.
+            * overwrite: Overwrite the row with the incoming data.
```
Comment (jreback) [inline pandas/core/generic.py:2816]: use 'replace' not overwrite

File: pandas/io/sql.py [hunk 3]
```diff
@@ -618,10 +619,19 @@ def to_sql(
     schema : str, optional
         Name of SQL schema in database to write to (if database flavor
         supports this). If None, use default schema (default).
-    if_exists : {'fail', 'replace', 'append'}, default 'fail'
+    if_exists : {'fail', 'replace', 'append'},
+        default 'fail'.
```
Comment (jreback) [inline pandas/io/sql.py:623]: try not to change things not-related it makes the diff that much bigger

File: pandas/core/generic.py [hunk 4]
```diff
@@ -2799,13 +2800,22 @@ def to_sql(
...
+        if_exists : {'fail', 'replace', 'append'},\
+        default 'fail'
             How to behave if the table already exists.
 
             * fail: Raise a ValueError.
             * replace: Drop the table before inserting new values.
             * append: Insert new values to the existing table.
 
+        on_row_conflict : {'fail', 'overwrite', 'ignore'},\
+        default 'fail'
```
Comment (jreback) [inline pandas/core/generic.py:2812]: needs a versionadded 1.5.0

File: pandas/io/sql.py [hunk 5]
```diff
@@ -671,9 +681,19 @@ def to_sql(
     `sqlite3 <https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.rowcount>`__ or
     `SQLAlchemy <https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.BaseCursorResult.rowcount>`__
     """  # noqa:E501
+
     if if_exists not in ("fail", "replace", "append"):
         raise ValueError(f"'{if_exists}' is not valid for if_exists")
 
+    if on_row_conflict not in ("fail", "overwrite", "ignore"):
+        raise ValueError(f"'{on_row_conflict}' is not valid for on_row_conflict'")
```
Comment (jreback) [inline pandas/io/sql.py:689]: is there a test for this?

File: pandas/io/sql.py [hunk 6]
```diff
@@ -671,9 +681,19 @@ def to_sql(
...
+    # on_row_conflict only used with append
```
Comment (jreback) [inline pandas/io/sql.py:690]: blank line before, comments can be *inside* the clause

File: pandas/io/sql.py [hunk 7]
```diff
@@ -827,6 +850,182 @@ def create(self):
         else:
             self._execute_create()
 
+    def _load_existing_pkeys(self, primary_keys, primary_key_values):
```
Comment (jreback) [inline pandas/io/sql.py:853]: type inputs & outputs

**Review by jreback — CHANGES_REQUESTED** [review]
i started reviewing, but this again is adding an enormous amount of complexity. can you get sqlalchemy to do this type of operation directly?

File: pandas/io/sql.py (hunk 5 above)
Comment (LSturtew) [reply pandas/io/sql.py:689]: yes:
```
    def test_to_sql_invalid_on_row_conflict(self, pkey_frame):
        msg = "'notvalidvalue' is not valid for on_row_conflict"
        with pytest.raises(ValueError, match=msg):
            sql.to_sql(
                pkey_frame,
                "pkey_frame1",
                self.conn,
                if_exists="append",
                on_row_conflict="notvalidvalue",
            )
```

**Comment by LSturtew** [comment]
Sqlalchemy handles row conflicts differently depending on the database dialect. E.g: for postgress and sqlite the implementation is similar to what was originally proposed in the original PR: `on_conflict_do_nothing` or `on_conflict_do_update`, for mysql on the other hand this is handled with `on_duplicate_key_update`.

**Comment by maveec** [comment]
@LSturtew seems little is missing to finalize this. Thank you for your effort and keep up the great work, lots of people are - and will be - grateful for this =)

**Comment by hichameyessou** [comment]
@LSturtew few months passed, do you still have the capacity to complete this PR?

**Comment by JustinGuese** [comment]
can we run this again? the logs for the tests are gone... Or what seems to be missing? I guess it is only a problem with the Docstrings and types, right?

**Comment by JustinGuese** [comment]
I have added a pr for some small docstring changes to the repo of @LSturtew, because if I create a new PR to the pandas main branch all the colab would be lost I guess?
Otherwise, if you (LSturtew) give editing rights to maintainers I can do it directly (https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

**Comment by LSturtew** [comment]
> https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

The box `Allow edits and access to secrets by maintainers` is checked

**Comment by JustinGuese** [comment]
okay, i can't get it to work... maybe I'm just stuck in my thoughts
```
ERROR: Permission to LSturtew/pandas.git denied to JustinGuese.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

I'm:
1. `gh pr checkout -R pandas-dev/pandas 46232`
2. adding my changes, git add & git commit
3. `git push git@github.com:LSturtew/pandas.git sql-upsert:sql-upsert`

what am I doing wrong?

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**